realm/realm-library/src/main/java/io/realm/internal/SharedGroup.java

```diff
@@ -293,6 +297,29 @@ public int hashCode() {
         }
     }
 
+    /**
+     * Waits for change committed by {@link SharedGroup} in other Thread.
+     *
+     * @return {@code true} if successfully detects change, {@code false} no change has detected otherwise.
+     */
+    public boolean waitForChange() {
+        if (isStoppedWaitingByUser) {
+            isStoppedWaitingByUser = false;
+            return false;
+        }
```

Comment:
@stk1m1 I think @beeender is correct about the timing here. If `stopWaitForChange` is executed after the `if`-statement and before calling `nativeWaitForChange()`, you will call `nativeStopWaitForChange()` before calling `nativeStartWaitForChange()` which can lead to the SharedGroup never stopping.

Making both methods `synchronized` should fix it from what I can tell.